File: backend/src/models/MCPServer.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IMCPTool {
    name: string;
    description?: string;
    inputSchema: Record<string, unknown>;
}

export interface IMCPHeader {
    key: string;
    value: string;
}

export interface IMCPServer extends Document {
    userId: mongoose.Types.ObjectId;
    name: string;
    description?: string;
    transport: 'http' | 'sse';
    url: string;
    headers: IMCPHeader[];
    enabled: boolean;
    cachedTools: IMCPTool[];
    lastConnectedAt?: Date;
    lastError?: string;
    createdAt: Date;
    updatedAt: Date;
}

const mcpToolSchema = new Schema<IMCPTool>(
    {
        name: { type: String, required: true },
        description: { type: String },
        inputSchema: {
            type: Schema.Types.Mixed,
            default: { type: 'object', properties: {} },
        },
    },
    { _id: false }
);

const mcpHeaderSchema = new Schema<IMCPHeader>(
    {
        key: { type: String, required: true, trim: true },
        value: { type: String, required: true },
    },
    { _id: false }
);

const mcpServerSchema = new Schema<IMCPServer>(
    {
        userId: {
            type: Schema.Types.ObjectId,
            ref: 'User',
            required: true,
            index: true,
        },
        name: {
            type: String,
            required: true,
            trim: true,
            maxlength: 80,
        },
        description: {
            type: String,
            trim: true,
            maxlength: 500,
        },
        transport: {
            type: String,
            enum: ['http', 'sse'],
            default: 'http',
        },
        url: {
            type: String,
            required: true,
            trim: true,
        },
        headers: {
            type: [mcpHeaderSchema],
            default: [],
        },
        enabled: {
            type: Boolean,
            default: true,
        },
        cachedTools: {
            type: [mcpToolSchema],
            default: [],
        },
        lastConnectedAt: {
            type: Date,
        },
        lastError: {
            type: String,
        },
    },
    {
        timestamps: true,
    }
);

mcpServerSchema.index({ userId: 1, name: 1 }, { unique: true });

export default mongoose.model<IMCPServer>('MCPServer', mcpServerSchema);

File: mobile/lib/data/models/mcp_server.dart
class MCPTool {
  final String name;
  final String? description;
  final Map<String, dynamic>? inputSchema;

  MCPTool({
    required this.name,
    this.description,
    this.inputSchema,
  });

  factory MCPTool.fromJson(Map<String, dynamic> json) {
    return MCPTool(
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      inputSchema: json['inputSchema'] as Map<String, dynamic>?,
    );
  }
}

class MCPServer {
  final String id;
  final String name;
  final String? description;
  final String transport; // 'http', 'sse'
  final String url;
  final List<Map<String, dynamic>> headers;
  final bool enabled;
  final List<MCPTool> cachedTools;
  final String? lastConnectedAt;
  final String? lastError;
  final String createdAt;
  final String updatedAt;

  MCPServer({
    required this.id,
    required this.name,
    this.description,
    required this.transport,
    required this.url,
    required this.headers,
    required this.enabled,
    required this.cachedTools,
    this.lastConnectedAt,
    this.lastError,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MCPServer.fromJson(Map<String, dynamic> json) {
    return MCPServer(
      id: json['_id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      transport: json['transport'] as String? ?? 'http',
      url: json['url'] as String? ?? '',
      headers: (json['headers'] as List<dynamic>?)
              ?.map((e) => e as Map<String, dynamic>)
              .toList() ??
          [],
      enabled: json['enabled'] as bool? ?? true,
      cachedTools: (json['cachedTools'] as List<dynamic>?)
              ?.map((e) => MCPTool.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      lastConnectedAt: json['lastConnectedAt'] as String?,
      lastError: json['lastError'] as String?,
      createdAt: json['createdAt'] as String? ?? '',
      updatedAt: json['updatedAt'] as String? ?? '',
    );
  }
}
